calculator/reducer.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  CalculatorActionKind,
+  CalculatorState,
+  calculatorDefaultState,
+  CalculatorAction,
+} from "./types";
+
+export const calculatorReducer = (
```

Comment:
Not a big fan of pulling in Reducer for something that's relatively simple. See my version to see how it can be done.

Reply:
I see! So, which situation is the best use case to adopt `useReducer` approach?

Reply:
Generally, i would stick with `State` and `Context` hooks.
The only time i would think `Reducer` is appropriate is the something like `State = S & T` where one component only needs `T` and the other needs `S`. However, there's a very specific action that actually updates both `S & T`.

Say I am in an app that look at analytics that's reported every year. Each report contains one part about sales data and the other part about inventory data. Each of these are fetched from a separate API.

When I change the **year** in the top level, both the current sales data and inventory data needs to be invalidated and re-fetch for the appropriate year. This is the case where `Reducer` would be required.

Because the alternative would be a poll-model, like

```ts
function useYear() {
   const {reload: reloadSales} = useSalesData();
   const {reload: reloadInventory} = useInventoryData();
   // .... which scales up 
   
   const changeYear = (year) => { reloadSales(year); reloadInventory(year); };
   return {
    changeYear,
  };
}
```

This isn't ideal because `useYear` shouldn't be aware of all the report data